File: backend/src/tour_guide/models/poi.py
from dataclasses import dataclass, field


@dataclass
class OsmNode:
    id: str  # e.g. "osm:node:12345"
    lat: float
    lon: float
    tags: dict[str, str] = field(default_factory=dict)


@dataclass
class WikiArticle:
    title: str
    extract: str  # intro text
    url: str
    lang: str


@dataclass
class POIContext:
    osm: OsmNode
    wiki: WikiArticle | None = None


@dataclass
class Place:
    id: str                          # "gplace:{place_id}"
    name: str
    lat: float
    lon: float
    rating: float | None
    user_ratings_total: int | None
    price_level: int | None          # 1-4, None if unknown
    types: list[str]
    vicinity: str


@dataclass
class POI:
    id: str
    name: str
    lat: float
    lon: float
    tags: dict[str, str] = field(default_factory=dict)
    wiki: WikiArticle | None = None
    distance_m: float = 0.0
    confidence: str = "low"          # "high" | "medium" | "low"
    # foodie-only fields (None for non-foodie POIs)
    rating: float | None = None
    user_ratings_total: int | None = None
    price_level: int | None = None
    place_types: list[str] | None = None
    vicinity: str | None = None


@dataclass
class BBox:
    min_lat: float
    min_lon: float
    max_lat: float
    max_lon: float


@dataclass
class TagFilter:
    key: str
    values: list[str] = field(default_factory=list)  # empty = any value

File: flutter_app/lib/shared/backend/models/poi.dart
class WikiArticle {
  final String title;
  final String extract;
  final String url;

  const WikiArticle({
    required this.title,
    required this.extract,
    required this.url,
  });

  factory WikiArticle.fromJson(Map<String, dynamic> json) => WikiArticle(
        title: json['title'] as String,
        extract: json['extract'] as String,
        url: json['url'] as String,
      );

  Map<String, dynamic> toJson() => {
        'title': title,
        'extract': extract,
        'url': url,
      };
}

class POI {
  final String id;
  final String name;
  final double lat;
  final double lon;
  final Map<String, String> tags;
  final WikiArticle? wiki;
  final double distanceM;
  final String confidence;

  // foodie only — null for non-foodie POIs
  final double? rating;
  final int? userRatingsTotal;
  final int? priceLevel;
  final List<String>? placeTypes;
  final String? vicinity;

  const POI({
    required this.id,
    required this.name,
    required this.lat,
    required this.lon,
    required this.tags,
    this.wiki,
    required this.distanceM,
    required this.confidence,
    this.rating,
    this.userRatingsTotal,
    this.priceLevel,
    this.placeTypes,
    this.vicinity,
  });

  factory POI.fromJson(Map<String, dynamic> json) => POI(
        id: json['id'] as String,
        name: json['name'] as String,
        lat: (json['lat'] as num).toDouble(),
        lon: (json['lon'] as num).toDouble(),
        tags: (json['tags'] as Map<String, dynamic>? ?? {})
            .cast<String, String>(),
        wiki: json['wiki'] != null
            ? WikiArticle.fromJson(json['wiki'] as Map<String, dynamic>)
            : null,
        distanceM: (json['distance_m'] as num).toDouble(),
        confidence: json['confidence'] as String,
        rating: (json['rating'] as num?)?.toDouble(),
        userRatingsTotal: json['user_ratings_total'] as int?,
        priceLevel: json['price_level'] as int?,
        placeTypes: (json['place_types'] as List<dynamic>?)?.cast<String>(),
        vicinity: json['vicinity'] as String?,
      );
}
